[3.5] Remove com_messages from core new installs

zero-24 opened the pull request:
Following the roadmap for Joomla 3.5: 
`Remove additional components and all associated pieces. ([...] com_messages).[...]`
http://developer.joomla.org/cms/roadmap.html

This will remove com_messages from the core new installs.
### how to test (first test)
- download my branche: https://github.com/zero-24/joomla-cms/archive/patch-13.zip
- install as normal joomla 3.4
- make sure com_messages is gone and no errors are there e.g. SQL or others.
### how to test (seccond test)
- install normal 3.4
- update with a update package (needs to be created and uploaded to joomla.org)
- see that nothing is changed and com_messages still work as expeced.
### Open questions
- what we need to do outsited of this PR?
- how can we create a update package and upload it to joomla.org? (for the seccond test)
### More information
- New com_messages repo: https://github.com/joomla-extensions/messages (i will open a PR to add the needed files to this repo. (based on com_weblinks repo)
- First PR is done more PRs against the new repo will follow.

zero-24 committed:
Delete messages.xml

zero-24 committed:
Delete messages.php

zero-24 committed:
Delete controller.php

zero-24 committed:
Delete config.xml

zero-24 committed:
Delete access.xml

zero-24 committed:
Delete default.php

zero-24 committed:
Delete view.html.php

zero-24 committed:
Delete default.php

zero-24 committed:
Delete edit.php

zero-24 committed:
Delete view.html.php

zero-24 committed:
Delete default.php

zero-24 committed:
Delete view.html.php

zero-24 committed:
Delete message.php

zero-24 committed:
Delete messages.php

zero-24 committed:
Delete message.php

zero-24 committed:
Delete config.php

zero-24 committed:
Delete config.xml

zero-24 committed:
Delete message.xml

zero-24 committed:
Delete usermessages.php

zero-24 committed:
Delete mysettings.php

zero-24 committed:
Delete messages.php

zero-24 committed:
Delete messages.php

zero-24 committed:
Delete config.php

zero-24 committed:
Delete message.php

zero-24 committed:
Delete messages.php

zero-24 committed:
Update joomla.sql

zero-24 committed:
Update joomla.sql

zero-24 committed:
Update joomla.sql

zero-24 committed:
Update joomla.sql

zero-24 committed:
Update joomla.sql

zero-24 committed:
Update mod_status.php

zero-24 committed:
Update default.php

zero-24 committed:
Update mod_status.php

zero-24 committed:
Update en-GB.mod_status.ini

dgt41 commented:
Great initiative!
I was hoping to remove the mootools modal for the options before decoupling it from the core. I guess I’ll have to make a PR to that repo then!

zero-24 commented:
@dgt41 i hope i can prepare a final PR against the new repo. So i think the easiest would be to send a PR against this repo if my PR is merged into it? I will create a issue on that repo.

dgt41 commented:
@zero-24 I had a PR #4563, which already got some tests etc, but I guess I can make a new one on the new repo. No need for issue.

zero-24 committed:
cs

```
FILE: ...build/joomla/joomla-cms/administrator/modules/mod_status/mod_status.php
--------------------------------------------------------------------------------
FOUND 3 ERROR(S) AFFECTING 3 LINE(S)
--------------------------------------------------------------------------------
 26 | ERROR | Whitespace found at end of line
 29 | ERROR | Whitespace found at end of line
 39 | ERROR | Whitespace found at end of line
--------------------------------------------------------------------------------
UPGRADE TO PHP_CODESNIFFER 2.0 TO FIX ERRORS AUTOMATICALLY
--------------------------------------------------------------------------------

```

https://travis-ci.org/joomla/joomla-cms/jobs/53394986

n9iels commented:
Good initiative!

I installed your Joomla! version with the Sample data: testing.
When you log in the first page you get, is a error page with this error:
```
 1146 Table 'joomla-cms-patch-13.c15is_messages_cfg' doesn't exist SQL=SELECT * FROM `c15is_messages_cfg` WHERE `user_id` = 101 AND `cfg_name` = 'auto_purge'
```
Looks like somewhere there is a query still seraching for the message table.
No father problems/com_message things found

zero-24 committed:
Update joomla.php

zero-24 commented:
Thanks @n9iels i will have a look into it.

zero-24 committed:
Update administrator.php

zero-24 commented:
@n9iels can you retest? This: https://github.com/zero-24/joomla-cms/commit/e9e1e9e47d22123927abd7159b09b15947aba22f should fix it

sovainfo commented:
What makes you say new content notification will be deprecated in J4?
Object to removing that functionality.

mbabker commented:
I think I'd suggest the feature be in a plugin distributed with the com_messages component versus being coupled to a core plugin that won't be remove.  Then that feature is deprecated in the core (as it's coupled to what should be a standalone extension) but available via a supported extension.

mbabker commented on administrator/language/en-GB/en-GB.mod_status.ini:
```diff
@@ -11,7 +11,7 @@ MOD_STATUS_FIELD_SHOW_LOGGEDIN_USERS_ADMIN_DESC="Show the number of users logged
 MOD_STATUS_FIELD_SHOW_LOGGEDIN_USERS_DESC="Show the number of users logged-in for both Frontend and Backend."
 MOD_STATUS_FIELD_SHOW_LOGGEDIN_USERS_LABEL="Show Logged-in Users"
-MOD_STATUS_FIELD_SHOW_MESSAGES_DESC="Show the messages count for the current user's inbox."
+MOD_STATUS_FIELD_SHOW_MESSAGES_DESC="Show the messages count for the current user's inbox, If the messages component is installed."
```
"If" should be "if" (lowercase i)

mbabker commented on administrator/modules/mod_status/tmpl/default.php:
```diff
@@ -42,18 +42,21 @@
 
-//  Print the inbox message.
-if ($params->get('show_messages', 1))
+if (JComponentHelper::isInstalled('com_messages'))
```
Can make this `if (JComponentHelper::isInstalled('com_messages') && $params->get('show_messages', 1))`, saves one nested if (and actually the diff only becomes the changed if line)

mbabker commented on libraries/cms/application/administrator.php:
```diff
@@ -339,50 +340,56 @@ public function login($credentials, $options = array())
-	 * @since   3.2
+	 * @since       3.2
+	 *
+	 * @deprecated  4.0  As com_messages is decoupled from core (see also line 336)
```
This whole method should probably move to a plugin event (I think onAfterUserLogin might be appropriate).  Obviously the method has to stay until 4.0, but moving the logic to a plugin keeps this extension specific feature with the extension.

sovainfo replied:
That would have been the correct place if it had been a private method. Unfortunately it is declared as public. Moving it to a plugin will not do. That means the method should probably move to JUser.

zero-24 replied:
The plugin is implemented in the new repo see: https://github.com/joomla-extensions/messages/blob/master/src/plg_system_messages/messages.php

should we place a logger / deprecated message here?

N6REJ replied:
this is the component that sends out system email notifications is it not?
Bear
On 6/30/2015 01:31, zero-24 wrote:
>
> In libraries/cms/application/administrator.php 
> <https://github.com/joomla/joomla-cms/pull/6352#discussion_r33543654>:
>
> >  	 *
> > -	 * @since   3.2
> > +	 * @since       3.2
> > +	 *
> > +	 * @deprecated  4.0  As com_messages is decoupled from core (see also line 336)
>
> The plugin is implemented in the new repo see: 
> https://github.com/joomla-extensions/messages/blob/master/src/plg_system_messages/messages.php
>
> should we place a logger / deprecated message here?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/joomla/joomla-cms/pull/6352/files#r33543654>.
>
> No virus found in this message.
> Checked by AVG - www.avg.com <http://www.avg.com>
> Version: 2015.0.6037 / Virus Database: 4365/10128 - Release Date: 06/30/15
>

zero-24 replied:
yes and no :smile: 

It should not but it is :cry: 

https://github.com/joomla/joomla-cms/blob/staging/plugins/content/joomla/joomla.php#L68-94

mbabker commented on plugins/content/joomla/joomla.php:
```diff
@@ -29,68 +29,78 @@ class PlgContentJoomla extends JPlugin
-		// Check we are handling the frontend edit form.
-		if ($context != 'com_content.form')
+		// This function don't work if com_messages ist not installed
+		if (JComponentHelper::isInstalled('com_messages'))
```
Just change this block to:

```php
if (!JComponentHelper::isInstalled('com_messages'))
{
    return true;
}
```

sovainfo commented:
Object to calling something deprecated as of J4. That sounds if the core supported component messages won't make it to J4. If that is already decided, that should be communicated properly. Moving functionality is ok, it should be part of this exercise. Removing it from the core and adding it to the messages project in core supported.
Moving and deprecating are two different things!

mbabker commented:
So what do you call a feature that is moved out of core API to another location?  By definition, if we're removing a feature from a class, then it's deprecated (even if it's just being moved to another location); support for a specific class and/or method API endpoint is scheduled for removal.  Using my suggestion on `JApplicationAdministrator::purgeMessages()`, we'd deprecate the method call and move what it does to another location.  This communicates that the API endpoint will no longer be supported in a future release but not necessarily that the feature is being fully removed.

sovainfo commented:
Consider it not applying to pieces of code of the plugin. The purpose of deprecating is to warn other developers not to use those methods. That doesn't apply here. With the move on messages a new content plugin or any other technical solution should be provided to maintain the functionality. 

Same applies to purgeMessages, its content should be changed to cater for the changes with messages component. The method should not be deprecated. It should start to provide the functionality depending on installation of messages and the possible replacement!

Because that is what moving these components to core supported means! Continue to use them or use an alternative. Part missing is what do I need to do to replace notification with something like uddeIM for example.

mbabker commented:
Personally, I think having our top level APIs bound to specific components or plugins is a bad idea.  Just on a glance, I see hardcoded references to com_languages, com_messages, com_users, the language filter plugin, and the database session storage; this couples our application to these components and makes it more complex to use an alternative user management system or database storage engine, for example.  Our top level APIs (such as the application classes) should be able to swap out implementation details (which is what the code in purgeMessages is, an implementation detail) with ease; if I want to use APC session management instead of my Joomla database, I should be able to.

In the case of purgeMessages, if we were to refactor it into a more abstract API that isn't bound to a specific extension or implementation, I'd love to see it; but in its current form with it being bound to only the messages component, it should be deprecated.

I think we actually agree on some parts of this, we're just looking at it from different perspectives.

zero-24 committed:
Update en-GB.mod_status.ini

zero-24 committed:
Update default.php

zero-24 committed:
Update joomla.php

zero-24 committed:
Update mod_status.php

zero-24 commented:
Thanks @mbabker your comments are fixed. regarding ` JApplicationAdministrator::purgeMessages()` I will implemet it via plugin into my PR against the new repo same for the mod_status code. Please let me some time to do it. :smile:

Worti2 commented:
Hi zero-24,
I made the first test, but there are still the language files:
/language/en-GB/en-GB.com-messages.ini
/administrator/language/en-GB.com-messages.ini
/administrator/language/en-GB.com-messages.sys.ini
They have to be removed, too, and the test has to be repeated.
In the database were no tables installed.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6352">issues.joomla.org/joomla-cms/6352</a>.</sub>

jehacgn commented:
system test für com_messages noch entfernen<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6352">issues.joomla.org/joomla-cms/6352</a>.</sub>

brianteeman commented:
As with weblinks the language files will not be removed

zero-24 commented:
@jehacgn  can you point me to the system tests?

dgt41 commented:
Guys can we apply first #4563 and then go on with this? The only thing that will @zero-24 have to change is to NOT to remove this file: `administrator/components/com_messages/layouts/toolbar/mysettings.php` as that will be removed in the other PR. Thanks!

zero-24 committed:
Create mysettings.php

zero-24 commented:
done @dgt41 https://github.com/joomla/joomla-cms/commit/2a5efed0182cd7f379cf06d7015b9d39c7e74f96

zero-24 committed:
Delete MessagingEditPage.php

zero-24 commented:
>system test für com_messages noch entfernen

Thanks @jehacgn just removed the test file https://github.com/zero-24/joomla-cms/commit/f0f3ea82f6832049729b4add4b2e1f70cebe7225

zero-24 commented:
@Worti2 and @jehacgn can you retest the last changes? The language files still there as same for com_weblinks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6352">issues.joomla.org/joomla-cms/6352</a>.</sub>

Worti2 commented:
Hi zero-24,
tested it and everything looks fine.
Greetings from the Möhnesee

zero-24 committed:
Delete edit.php

zero-24 committed:
Delete default.php

zero-24 committed:
lets try to fix it

zero-24 committed:
recommit the update sql files

zero-24 commented:
Please notice this: https://github.com/joomla/joomla-cms/pull/6352#discussion_r33571025

com_message is used for messaging e.g. if you have a new article.
https://github.com/joomla/joomla-cms/blob/staging/plugins/content/joomla/joomla.php#L68-94

So should we implement a other solution to implement it outside of com_messages or simple not remove messages?

zero-24 committed:
fix it

smz commented:
@zero-24
off the top of my head, I'd just get rid of those lines in https://github.com/joomla/joomla-cms/blob/staging/plugins/content/joomla/joomla.php#L68-94 and add an (optional) plugin acting on ContentAfterSave in the decoupled com_messages.

3rd party extension could implement similar solutions for different messaging systems.

Is this feasible?

infograf768 commented:
Suggestions: the check should not only be for the component being installed, but also to be enabled.
something like
```
if (!JComponentHelper::getComponent('com_messages', true)->enabled)
```

Also, it seems that mod_status would need an `onContentPrepareForm` to not display the `email_new_fe` field when the component is not installed.

That is if we do not add a new specific plugin.
I wonder how B/C that would be.

Bakual commented:
> Suggestions: the check should not only be for the component being installed, but also to be enabled.
something like `if (!JComponentHelper::getComponent('com_messages', true)->enabled)`

There is a helper method `JComponentHelper::isEnabled('com_messages')` :smile:

infograf768 commented:
Looking further in JComponentHelper, it looks to me that both methods would not be enough as "enabled" is not checked at all.

To test, just edit the _extensions table for com_messages and set enabled to 0 (the component status can't be edited from back-end as it is protected)
Then add in Isis index.php:
`<?php var_dump(JComponentHelper::isEnabled('com_messages')); ?>`
or
`<?php var_dump(JComponentHelper::getComponent('com_messages', true)->enabled); ?>`

You will get 1 instead of 0.
This is due to the function isEnabled() which only uses the getComponent() method to set its result to enabled:
```
	public static function isEnabled($option)
	{
		$result = static::getComponent($option, true);

		return $result->enabled;
	}
```

`isInstalled() does only checks for the existence of the component in the db whatever its status.

Looks like our code is missing something here as the notion of "enabled" in the helper does not correspond to the "enabled" status column in _extensions.

infograf768 commented:
My mistake. forget it. deleting above comment.

zero-24 commented:
@infograf768 @Bakual 

>There is a helper method JComponentHelper::isEnabled('com_messages') :smile:

That breaks B/C

The "is installed check" is allreay implementet. See: https://github.com/joomla/joomla-cms/pull/6352/files#diff-4bec224e62bf4c1a53784d49773e3841R38

But if we check that messages is enabled we break all sites that currently has disabled messages but use the notifications.

And this means a B/C break.

Install 3.4.3 --> Messaging on submiting a article works out of the box
Install 3.5.0 (or other version without messages) --> Messaging on submiting a article don't works.

Bakual commented:
How does messaging work when com_messages is disabled? I would call that a bug. Or do I miss something?

I'm fine however with keeping existing behavior for now.

zero-24 commented:
Lets call it a hack ;)
https://github.com/joomla/joomla-cms/blob/staging/plugins/content/joomla/joomla.php#L68-94

Currently the plugin don't care if com_messages is enabled or not. It includes the files and send the messages out.

Bakual commented:
I'd call that a bug, not a hack :stuck_out_tongue:

zero-24 commented:
But if we change the way we handel that messages we break B/C anyway so this needs to wait for 4.0 but if the PLT decides to do this with 3.5 and disable the mail sending with com_messages this PR can be merged ;)

Bakual commented:
Honestly? It's a bug which should be fixed. And chances are low that somebody disabled com_messages. Especially since it's protected and can't be disabled without hacking the database :smile:

infograf768 commented:
>And chances are low that somebody disabled com_messages. Especially since it's protected and can't be disabled without hacking the database

Agree.

@zero-24
Also, please do not forget
>Also, it seems that mod_status would need an onContentPrepareForm to not display the email_new_fe field when the component is not installed.

zero-24 commented:
@infograf768 

>Also, it seems that mod_status would need an onContentPrepareForm to not display the email_new_fe field when the component is not installed.

I'm not sure how to do this?

@Bakual 

>Especially since it's protected and can't be disabled without hacking the database

Oh. I don't know. This way it can't be uninstalled and i think should not decoupled.

Personaly I think we should close here and dont decouple it with 3.5 and wait for 4.0 to decouple it. We have much more easy components to decouple. And have some issues on decoupeling (e.g. the SSL issue and the others that was show by the PLT). If we decouple this extension we break B/C and have one more extension that makes the same problems as weblinks.

If someone what to take the work I will leave my branch. But personaly i think we should wait with messages for 4.0 and first fixes the issues that was pointed out by the PLT.

Bakual commented:
@zero-24 This one imho would be fine to decouple. It is currently protected because there are issues if you uninstall it. Exactly those we discussed here.
There is no B/C with this PR and Weblinks has no issues beside the SSL one which isn't a weblinks issue.

infograf768 commented:
For `email_new_fe` , I meant plugins/content/joomla/joomla.php
In mod_status we need to deal with `show_messages`

Kubik-Rubik commented:
I understand what @infograf768 means. Those fields (options) should not be visible in the settings if the component is not installed and enabled. 

@Bakual Do you have an idea how to the the checks? Maybe a new extra field for the check with the combination of "showon"?

Bakual commented:
`showon` would be to complicate. Just create a new custom field which extends from `JFormFieldRadio` where you add the check. If it fails, you could either show a message indicating com_messages is disabled/uninstalled, make the field readonly or you just hide the field completely. Just pick something that makes sense :)

asika32764 commented:
> Personaly I think we should close here and dont decouple it with 3.5 and wait for 4.0 to decouple it. We have much more easy components to decouple. And have some issues on decoupeling (e.g. the SSL issue and the others that was show by the PLT). If we decouple this extension we break B/C and have one more extension that makes the same problems as weblinks.

Could we just unprotected this component that I can remove it in 3.5?

zero-24 commented:
>Could we just unprotected this component that I can remove it in 3.5?

Yes we can. But i still don't think this is the best extension to start with ;)

And we change the behavior for all new installs that I recommend to do with Joomla 4.0 and not with 3.5. If we want to do it B/C we also need much "is installed" / "is enabled" checks.

And something very special (i don't know to implement) for mod_status as noted above by JM.

Feel free to fork: https://github.com/zero-24/joomla-cms/tree/patch-13

infograf768 commented:
@zero-24
Bakual's  proposal is indeed the simplest. We just have to add in the module and in the plugin a new field type and change in the xml to this new field type.
Somethng like this:
```
<?php
/**
 * @package     Joomla.Platform
 * @subpackage  Form
 *
 * @copyright   Copyright (C) 2005 - 2015 Open Source Matters, Inc. All rights reserved.
 * @license     GNU General Public License version 2 or later; see LICENSE
 */

defined('JPATH_PLATFORM') or die;

/**
 * Form Field class for the Joomla Platform.
 * Provides radio button inputs checking if com_messages is enabled
 *
 * @link   http://www.w3.org/TR/html-markup/command.radio.html#command.radio
 * @since  11.1
 */
class JFormFieldRadioMessages extends JFormFieldRadio
{
	/**
	 * The form field type.
	 *
	 * @var    string
	 * @since  11.1
	 */
	protected $type = 'RadioMessages';

	/**
	 * Method to get the radio button field input markup.
	 *
	 * @return  string  The field input markup.
	 *
	 * @since   11.1
	 */
	protected function getInput()
	{
		if (JComponentHelper::isEnabled('com_messages'))
		{
			return parent::getInput();
		}
		else
		{
			return false;
		}
	}
	
	/**
	 * Method to get the field label markup.
	 *
	 * @return  string  The field label markup.
	 *
	 * @since   3.4
	 */
	protected function getLabel()
	{
		if (JComponentHelper::isEnabled('com_messages'))
		{
			return parent::getLabel();
		}
		else
		{
			return false;
		}
	}
}
```
